### PR TITLE
Refactor low memory test into generic fault injection

### PR DESCRIPTION
**Pull request by Alan-Jowett**
## Description

To make it easier to implement #2101, this PR refactors and simplifies the fault injection logic. It exposes a simple set of APIs that modules can use to perform deterministic fault injection.

## Testing

CI/CD. Existing low memory tests.

## Documentation

No.

### CODE REVIEW

**Comment by Alan-Jowett** [comment]
@dthaler this is blocked on "low_memory" tests, which I ended up renaming to "fault_injection". Can we get the required tests switched over / suspended until this is merged?

**Comment by dthaler** [comment]
> @dthaler this is blocked on "low_memory" tests, which I ended up renaming to "fault_injection". Can we get the required tests switched over / suspended until this is merged?

Done

File: libs/platform/user/ebpf_fault_injection.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#endif
+
+    /**
+     * @brief Initialize fault injection.  This must be called before any other
+     * fault injection functions. This function is not thread safe.
+     *
+     * @param[in] stack_depth Number of stack frames to capture when a fault is
+     * injected.
+     * @retval ebpf_result_t EBPF_SUCCESS if successful.
+     * @retval EBPF_NO_MEMORY if memory allocation fails.
```
Comment (dthaler) [inline libs/platform/user/ebpf_fault_injection.h:19]:
```suggestion
     * @retval EBPF_SUCCESS The operation was successful.
     * @retval EBPF_NO_MEMORY Memory allocation failed.
```

File: libs/platform/user/ebpf_fault_injection.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     * fault injection functions. This function is not thread safe.
+     */
+    void
+    ebpf_fault_injection_uninitialize() noexcept;
+
+    /**
+     * @brief Enable fault injection. This function is thread safe.
+     *
+     * @return true Fault should be injected.
+     * @return false Fault should not be injected.
```
Comment (dthaler) [inline libs/platform/user/ebpf_fault_injection.h:35]:
```suggestion
     * @retval true Fault should be injected.
     * @retval false Fault should not be injected.
```

File: libs/platform/user/ebpf_fault_injection.cpp [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+  public:
+    /**
+     * @brief Construct a new ebpf fault injection object.
+     * @param[in] stack_depth The number of stack frames to compare when tracking faults.
+     */
+    _ebpf_fault_injection(size_t stack_depth);
+
+    /**
+     * @brief Destroy the ebpf fault injection object.
+     *
```
Comment (dthaler) [inline libs/platform/user/ebpf_fault_injection.cpp:33]:
```suggestion
```

File: libs/platform/user/ebpf_fault_injection.cpp [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+ */
+class ebpf_fault_injection_recursion_guard
+{
+  public:
+    ebpf_fault_injection_recursion_guard() { _ebpf_fault_injection_recursion++; }
+    ~ebpf_fault_injection_recursion_guard() { _ebpf_fault_injection_recursion--; }
+    /**
+     * @brief Return true if the current thread is recursing from the fault injection callback.
+     * @retval true
+     * @retval false
```
Comment (dthaler) [inline libs/platform/user/ebpf_fault_injection.cpp:149]:
```suggestion
     * @retval true The current thread is recursing from the fault injection callback.
     * @retval false The current thread is not recursing from the fault injection callback.
```

File: libs/platform/user/ebpf_fault_injection.cpp [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+    unsigned long hash;
+    // Capture EBPF_FAULT_STACK_CAPTURE_FRAME_COUNT_FOR_HASH frames of the current stack trace.
+    // The first EBPF_FAULT_STACK_CAPTURE_FRAMES_TO_SKIP frames are skipped to avoid
+    // capturing the fault injection code.
+    if (CaptureStackBackTrace(
+            EBPF_FAULT_STACK_CAPTURE_FRAMES_TO_SKIP,
+            static_cast<unsigned int>(stack.size()),
+            reinterpret_cast<void**>(stack.data()),
+            &hash) > 0) {
+        // Form the canonical stack
```
Comment (dthaler) [inline libs/platform/user/ebpf_fault_injection.cpp:198]:
```suggestion
        // Form the canonical stack.
```

**Review by shpalani — APPROVED** [review]

**Review by gtrevi — APPROVED** [review]

**Pull request merged**